[Remote Inspection] Add support for targets in nested shadow roots

whsieh opened the pull request:
#### c318bffec39ae54ee6bcfefaf129d1ab52790bde
<pre>
[Remote Inspection] Add support for targets in nested shadow roots
<a href="https://bugs.webkit.org/show_bug.cgi?id=273076">https://bugs.webkit.org/show_bug.cgi?id=273076</a>
<a href="https://rdar.apple.com/123160948">rdar://123160948</a>

Reviewed by Tim Horton, Abrar Protyasha and Aditya Keerthi.

Currently, `visibilityAdjustmentSelectors` is a flattened list of CSS selectors, which correspond to
an arbitrary number of targeted elements. This has a couple of drawbacks:

1.  We might resolve more targets than there originally were (in the case where a targeted element
    had multiple potential selectors, but one of them now maps to something else).

2.  It&apos;s unable to resolve targeted elements inside of shadow roots.

I previously worked around this in 277171@main by just returning an empty list in the case where the
element is inside of a shadow tree, but that only fixed the case where the targeted element itself
contained a shadow root (rather than being inside of the shadow DOM). This patch actually fixes this
issue, by revamping how we both surface and ingest CSS selectors corresponding to targeted elements:

•   `TargetedElementInfo` now contains a array of array of selectors. Each element in the outer
    array corresponds to an enclosing shadow host element that contains the targeted element (such
    that the final array maps to the targeted element itself). For targets that are not inside
    shadow roots, this simply means that there&apos;s a single outer array. The inner array, like before,
    represents a array of selectors (sorted by length) that can be used to resolve the target (or
    shadow host containing the target) within its tree scope.

•   `WKWebpagePreferences` now has a `_visibilityAdjustmentSelectorsIncludingShadowHosts` property,
    which consists of an array of array of sets of strings. To unpack this:

    1.  Each array in the outermost (first) array represents a single targeted element.

    2.  Each set in the second array represents a targeted element or shadow host containing the
        targeted element. The last element is the targeted element itself, and the first element (if
        the targeted element is inside of a shadow root) is the outermost shadow host element in the
        document.

    3.  Each string in the set represents a CSS selector that can potentially be used to resolve the
        targeted element or one of its hosts, within its tree scope.

To make this work, we:

(a) Make `TargetedElementController` recursively ascend the DOM by shadow host, to construct a
    `Vector&lt;Vector&lt;String&gt;&gt;` for each targeted element.

(b) Teach `TargetedElementController` to ingest the array of array of sets described above to
    resolve targeted elements that reside inside shadow trees.

There is no change in behavior for the case where targeted elements aren&apos;t inside the shadow DOM.

Test: ElementTargeting.AdjustVisibilityForTargetsInShadowRoot

* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::visibilityAdjustmentSelectors const):
(WebCore::DocumentLoader::setVisibilityAdjustmentSelectors):
* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::querySelectorMatchesOneElement):
(WebCore::computeIDSelector):
(WebCore::selectorsForTarget):

For elements inside of shadow roots, teach this to recursively find selectors that resolve to shadow
hosts, and return a list of lists of selectors.

(WebCore::targetedElementInfo):
(WebCore::ElementTargetingController::applyVisibilityAdjustmentFromSelectors):

Teach this to apply visibility adjustment to content inside of shadow roots.

* Source/WebCore/page/ElementTargetingController.h:
* Source/WebCore/page/ElementTargetingTypes.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Turn the list of selectors for each targeted element into a list of lists instead. See comments
above for more detail.

* Source/WebKit/Shared/WebsitePoliciesData.h:
* Source/WebKit/Shared/WebsitePoliciesData.serialization.in:

Turn the set of selectors into a list of list of sets. See comments above for more detail.

* Source/WebKit/UIProcess/API/APITargetedElementInfo.h:
* Source/WebKit/UIProcess/API/APIWebsitePolicies.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
(-[WKWebpagePreferences _setVisibilityAdjustmentSelectorsIncludingShadowHosts:]):
(-[WKWebpagePreferences _visibilityAdjustmentSelectorsIncludingShadowHosts]):
(-[WKWebpagePreferences _setVisibilityAdjustmentSelectors:]):
(-[WKWebpagePreferences _visibilityAdjustmentSelectors]):

Implement these on top of the new `_visibilityAdjustmentSelectorsIncludingShadowHosts` property.
Once no internal client relies on this anymore, we can simply remove it.

* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferencesPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTargetedElementInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTargetedElementInfo.mm:
(-[_WKTargetedElementInfo selectors]):
(-[_WKTargetedElementInfo selectorsIncludingShadowHosts]):
(-[_WKTargetedElementInfo isInShadowTree]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
(TestWebKitAPI::TEST(ElementTargeting, AdjustVisibilityFromPseudoSelectors)):
(TestWebKitAPI::TEST(ElementTargeting, AdjustVisibilityForTargetsInShadowRoot)):
(TestWebKitAPI::TEST(ElementTargeting, TargetContainsShadowRoot)):
(TestWebKitAPI::TEST(ElementTargeting, ContentInsideShadowRoot)): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/element-targeting-8.html: Added.

Add an API test to exercise all three of the new SPIs. Also rename `ContentInsideShadowRoot` to
`TargetContainsShadowRoot`, the latter of which is a more accurate name (since the targeted element
itself isn&apos;t inside the shadow DOM).

Canonical link: <a href="https://commits.webkit.org/277907@main">https://commits.webkit.org/277907@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/45187dfbbbd6cba1fd6884e2349ccb6643d2d4c0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48879 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28092 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51846 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51566 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44945 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51184 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34044 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25620 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39959 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49461 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25744 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42148 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21062 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23205 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43322 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6934 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45137 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43820 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53477 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23930 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20196 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47268 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25193 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42357 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46220 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10774 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26002 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24913 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -978,58 +996,95 @@ void ElementTargetingController::applyVisibilityAdjustmentFromSelectors(Document
-            continue;
+        bool foundLastTarget = false;
+        Ref<ContainerNode> containerToQuery = document;
+        for (size_t index = 0; index < selectorsForElementIncludingShadowHosts.size(); ++index) {
```
It seems like we could use an iterator here rather than an index based loop?

whsieh replied:
Ah, yeah — I only used an index here to know when I've resolved the final target, but I can track that index separately, and use a iterator.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
```diff
@@ -656,20 +656,61 @@ - (void)_setNetworkConnectionIntegrityPolicy:(_WKWebsiteNetworkConnectionIntegri
+            continue;
+        }
+
+        for (auto& selector : elementSelectors[0])
```
```suggestion
        for (auto& selector : elementSelectors.first())
```

whsieh committed:
[Remote Inspection] Add support for targets in nested shadow roots
https://bugs.webkit.org/show_bug.cgi?id=273076
rdar://123160948

Reviewed by Tim Horton, Abrar Protyasha and Aditya Keerthi.

Currently, `visibilityAdjustmentSelectors` is a flattened list of CSS selectors, which correspond to
an arbitrary number of targeted elements. This has a couple of drawbacks:

1.  We might resolve more targets than there originally were (in the case where a targeted element
    had multiple potential selectors, but one of them now maps to something else).

2.  It's unable to resolve targeted elements inside of shadow roots.

I previously worked around this in 277171@main by just returning an empty list in the case where the
element is inside of a shadow tree, but that only fixed the case where the targeted element itself
contained a shadow root (rather than being inside of the shadow DOM). This patch actually fixes this
issue, by revamping how we both surface and ingest CSS selectors corresponding to targeted elements:

•   `TargetedElementInfo` now contains a array of array of selectors. Each element in the outer
    array corresponds to an enclosing shadow host element that contains the targeted element (such
    that the final array maps to the targeted element itself). For targets that are not inside
    shadow roots, this simply means that there's a single outer array. The inner array, like before,
    represents a array of selectors (sorted by length) that can be used to resolve the target (or
    shadow host containing the target) within its tree scope.

•   `WKWebpagePreferences` now has a `_visibilityAdjustmentSelectorsIncludingShadowHosts` property,
    which consists of an array of array of sets of strings. To unpack this:

    1.  Each array in the outermost (first) array represents a single targeted element.

    2.  Each set in the second array represents a targeted element or shadow host containing the
        targeted element. The last element is the targeted element itself, and the first element (if
        the targeted element is inside of a shadow root) is the outermost shadow host element in the
        document.

    3.  Each string in the set represents a CSS selector that can potentially be used to resolve the
        targeted element or one of its hosts, within its tree scope.

To make this work, we:

(a) Make `TargetedElementController` recursively ascend the DOM by shadow host, to construct a
    `Vector<Vector<String>>` for each targeted element.

(b) Teach `TargetedElementController` to ingest the array of array of sets described above to
    resolve targeted elements that reside inside shadow trees.

There is no change in behavior for the case where targeted elements aren't inside the shadow DOM.

Test: ElementTargeting.AdjustVisibilityForTargetsInShadowRoot

* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::visibilityAdjustmentSelectors const):
(WebCore::DocumentLoader::setVisibilityAdjustmentSelectors):
* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::querySelectorMatchesOneElement):
(WebCore::computeIDSelector):
(WebCore::selectorsForTarget):

For elements inside of shadow roots, teach this to recursively find selectors that resolve to shadow
hosts, and return a list of lists of selectors.

(WebCore::targetedElementInfo):
(WebCore::ElementTargetingController::applyVisibilityAdjustmentFromSelectors):

Teach this to apply visibility adjustment to content inside of shadow roots.

* Source/WebCore/page/ElementTargetingController.h:
* Source/WebCore/page/ElementTargetingTypes.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Turn the list of selectors for each targeted element into a list of lists instead. See comments
above for more detail.

* Source/WebKit/Shared/WebsitePoliciesData.h:
* Source/WebKit/Shared/WebsitePoliciesData.serialization.in:

Turn the set of selectors into a list of list of sets. See comments above for more detail.

* Source/WebKit/UIProcess/API/APITargetedElementInfo.h:
* Source/WebKit/UIProcess/API/APIWebsitePolicies.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
(-[WKWebpagePreferences _setVisibilityAdjustmentSelectorsIncludingShadowHosts:]):
(-[WKWebpagePreferences _visibilityAdjustmentSelectorsIncludingShadowHosts]):
(-[WKWebpagePreferences _setVisibilityAdjustmentSelectors:]):
(-[WKWebpagePreferences _visibilityAdjustmentSelectors]):

Implement these on top of the new `_visibilityAdjustmentSelectorsIncludingShadowHosts` property.
Once no internal client relies on this anymore, we can simply remove it.

* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferencesPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTargetedElementInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTargetedElementInfo.mm:
(-[_WKTargetedElementInfo selectors]):
(-[_WKTargetedElementInfo selectorsIncludingShadowHosts]):
(-[_WKTargetedElementInfo isInShadowTree]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
(TestWebKitAPI::TEST(ElementTargeting, AdjustVisibilityFromPseudoSelectors)):
(TestWebKitAPI::TEST(ElementTargeting, AdjustVisibilityForTargetsInShadowRoot)):
(TestWebKitAPI::TEST(ElementTargeting, TargetContainsShadowRoot)):
(TestWebKitAPI::TEST(ElementTargeting, ContentInsideShadowRoot)): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/element-targeting-8.html: Added.

Add an API test to exercise all three of the new SPIs. Also rename `ContentInsideShadowRoot` to
`TargetContainsShadowRoot`, the latter of which is a more accurate name (since the targeted element
itself isn't inside the shadow DOM).

Canonical link: https://commits.webkit.org/277907@main
